File: C++ General/CANifier Demo/src/main/cpp/Tasks/TaskPWMmotorController.cpp
#include "Tasks/TaskPWMmotorController.h"
#include "ctre/Phoenix.h"
#include "Platform/Platform.h"

/* Destructor */
TaskPWMmotorController::~TaskPWMmotorController() { }

/* ILoopable */
void TaskPWMmotorController::OnStart() {
	/* If already running, do nothing */
	if (_running)
		return;

	/* Start transmitting neutral signal */
	_percentOutput = 0;
	Hardware::canifier->SetPWMOutput(Constants::MotorControllerCh, 0);
	Hardware::canifier->EnablePWMOutput(Constants::MotorControllerCh, true);

	/* Task is now running */
	_running = true;
}
void TaskPWMmotorController::OnStop() {
	/* Stop transmitting PWM */
	Hardware::canifier->EnablePWMOutput(Constants::MotorControllerCh, false);

	/* Task has stopped, now take note */
	_running = false;
}
bool TaskPWMmotorController::IsDone() {
	return false;
}
void TaskPWMmotorController::OnLoop() {
	/* Grab the axis and direct control the components */
	float axis = Hardware::gamepad->GetRawAxis(Constants::Gamepad_y);
	/* Scale to typical PWM widths */
	float pulseUs = LinearInterpolation::Calculate(axis, -1, 1000, +1, 2000); /* [-1,+1] => [1000,2000]us */
	/* Scale to period */
	float periodUs = 4200; // hard-coded for now, this will be settable in future firmware update.
	_percentOutput = pulseUs / periodUs;
	/* Set it */
	Hardware::canifier->SetPWMOutput(Constants::MotorControllerCh, _percentOutput);
}

File: Java General/CANifier Demo/src/main/java/frc/robot/Tasks/TaskPWMmotorController.java
/**
 * TaskPWMmotorController Class
 * 
 * ILoopable Task for using a joystick to drive a PWM motor controller.
 * Demonstrate CANifier's ability to generate PWM output for peripherals.
 */
package frc.robot.Tasks;

import frc.robot.Framework.LinearInterpolation;
import frc.robot.Platform.Constants;
import frc.robot.Platform.Hardware;
import com.ctre.phoenix.ILoopable;

public class TaskPWMmotorController implements ILoopable {
	float _percentOut;
	boolean _running; /* Assists TaskMainLoop with tracking, used within init */

	/* ILoopable */
	public void onStart() {
		/* If we are already running, nothing to do */
		if (_running) { return; }

		/* Start transmitting neutral */
		_percentOut = 0;
		Hardware.canifier.setPWMOutput(Constants.kMotorControllerCh.value, 0);
		Hardware.canifier.enablePWMOutput(Constants.kMotorControllerCh.value, true);

		/* Task is now running */
		_running = true;
	}

	public void onStop() {
		/* Stop transmitting PWM */
		Hardware.canifier.enablePWMOutput(Constants.kMotorControllerCh.value, false);

		/* Task has stopped */
		_running = false;
	}

	public boolean isDone() { return false; }

	public void onLoop() {
		/* Grab three axis and direct control the PWM MotorController */
		float axis = (float) Hardware.gamepad.getRawAxis(Constants.GamePadAxis_y);
		/* Scale to typical PWM widths ([-1,+1] => [1000,2000]us)*/
        float pulseUs = LinearInterpolation.calculate(axis, -1, 1000f, +1, 2000f);
		/* Scale to period */
		float periodUs = 4200; // Hard-coded for now, will be settable in future update
		_percentOut = pulseUs / periodUs;
		/* Set PWM Motor Controller */
		Hardware.canifier.setPWMOutput(Constants.kMotorControllerCh.value, _percentOut);
	}

	public String toString() {
		return "TaskPWMmotorController:";
	}
}
